Wrong struct name after #352 merged

Before #352, the struct name is directly generated by quote! macro
https://github.com/paritytech/subxt/blob/934aebcc2caca1df1c3e30aec319f1eb1d2dd253/codegen/src/types/type_def.rs#L124

After #352, will create a `CompositeDef::struct_def` https://github.com/paritytech/subxt/blob/b0004ea79fee38f6186ad85ac8876bd9a7cb90ed/codegen/src/types/type_def.rs#L88

The question is the `sruct_def` function will change the name to camel case style https://github.com/paritytech/subxt/blob/b0004ea79fee38f6186ad85ac8876bd9a7cb90ed/codegen/src/types/composite_def.rs#L88


After doing this. if we have two upper case words name's struct. like this
https://github.com/darwinia-network/darwinia-common/blob/1861a487b2d15a4831367e2e3ff3e2a73e2e9c8a/frame/staking/src/slashing.rs#L84

The generated code will output some errors.

```
error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:3676:64
     |
3676 |                     runtime_types::darwinia_staking::slashing::RK<
     |                                                                ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`
...
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:3695:73
     |
3695 |                 type Value = runtime_types::darwinia_staking::slashing::RK<
     |                                                                         ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`
...
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:4245:68
     |
4245 |                         runtime_types::darwinia_staking::slashing::RK<
     |                                                                    ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`
...
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:4271:68
     |
4271 |                         runtime_types::darwinia_staking::slashing::RK<
     |                                                                    ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`
...
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:9018:77
     |
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here
...
9018 |                     pub slashed: runtime_types::darwinia_staking::slashing::RK<_0, _0>,
     |                                                                             ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:9019:78
     |
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here
...
9019 |                     pub paid_out: runtime_types::darwinia_staking::slashing::RK<_0, _0>,
     |                                                                              ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:9172:73
     |
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here
...
9172 |                     pub own: runtime_types::darwinia_staking::slashing::RK<_1, _1>,
     |                                                                         ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:9175:68
     |
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here
...
9175 |                         runtime_types::darwinia_staking::slashing::RK<_1, _1>,
     |                                                                    ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`

error[E0412]: cannot find type `RK` in module `runtime_types::darwinia_staking::slashing`
    --> src\runtime.rs:9178:76
     |
8999 |                 pub struct Rk<_0, _1> {
     |                 --------------------- similarly named struct `Rk` defined here
...
9178 |                     pub payout: runtime_types::darwinia_staking::slashing::RK<_1, _1>,
     |                                                                            ^^ help: a struct with a similar name exists (notice the capitalization): `Rk`

error[E0282]: type annotations needed
    --> src\runtime.rs:9017:39
     |
9017 |                 pub struct SpanRecord<_0, _1> {
     |                                       ^^ cannot infer type for type parameter `_0`

error[E0283]: type annotations needed
    --> src\runtime.rs:9168:21
     |
9168 |                     :: subxt :: codec :: Encode, :: subxt :: codec :: Decode, Debug, Clone,
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `_0`
     |
     = note: cannot satisfy `_0: Encode`
note: required by a bound in `Encode`
    --> d:/opt/scoop/persist/rustup/.cargo\registry\src\github.com-1ecc6299db9ec823\parity-scale-codec-2.3.1\src\codec.rs:208:1
     |
208  | / pub trait Encode {
209  | |     // !INTERNAL USE ONLY!
210  | |     // This const helps SCALE to optimize the encoding/decoding by doing fake specialization.
211  | |     #[doc(hidden)]
...    |
251  | |     }
252  | | }
     | |_^ required by this bound in `Encode`
     = note: this error originates in the derive macro `::subxt::codec::Encode` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0282]: type annotations needed
    --> src\runtime.rs:9170:43
     |
9170 |                 pub struct UnappliedSlash<_0, _1, _2> {
     |                                           ^^ cannot infer type for type parameter `_0`

Some errors have detailed explanations: E0282, E0283, E0412.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `pangoro-subxt` due to 12 previous errors
```

--- a/codegen/src/api/calls.rs
+++ b/codegen/src/api/calls.rs
@@ -43,7 +43,7 @@ pub fn generate_calls(
     let struct_defs = super::generate_structs_from_variants(
         type_gen,
         call.ty.id(),
-        |name| name.to_camel_case(),
+        |name| name.to_camel_case().into(),
         "Call",
     );
     let (call_structs, call_fns): (Vec<_>, Vec<_>) = struct_defs

--- a/codegen/src/api/calls.rs
+++ b/codegen/src/api/calls.rs
@@ -22,7 +22,10 @@ use frame_metadata::{
     PalletCallMetadata,
     PalletMetadata,
 };
-use heck::SnakeCase as _;
+use heck::{
+    CamelCase as _,
+    SnakeCase as _,
+};
 use proc_macro2::TokenStream as TokenStream2;
 use proc_macro_error::abort_call_site;
 use quote::{
@@ -37,8 +40,12 @@ pub fn generate_calls(
     call: &PalletCallMetadata<PortableForm>,
     types_mod_ident: &syn::Ident,
 ) -> TokenStream2 {
-    let struct_defs =
-        super::generate_structs_from_variants(type_gen, call.ty.id(), "Call");
+    let struct_defs = super::generate_structs_from_variants(
+        type_gen,
+        call.ty.id(),
+        |name| name.to_camel_case(),
+        "Call",
+    );
     let (call_structs, call_fns): (Vec<_>, Vec<_>) = struct_defs
         .iter()
         .map(|struct_def| {

--- a/codegen/src/api/events.rs
+++ b/codegen/src/api/events.rs
@@ -32,7 +32,7 @@ pub fn generate_events(
     let struct_defs = super::generate_structs_from_variants(
         type_gen,
         event.ty.id(),
-        |name| name.to_string(),
+        |name| name.into(),
         "Event",
     );
     let event_structs = struct_defs.iter().map(|struct_def| {

--- a/codegen/src/api/events.rs
+++ b/codegen/src/api/events.rs
@@ -29,8 +29,12 @@ pub fn generate_events(
     event: &PalletEventMetadata<PortableForm>,
     types_mod_ident: &syn::Ident,
 ) -> TokenStream2 {
-    let struct_defs =
-        super::generate_structs_from_variants(type_gen, event.ty.id(), "Event");
+    let struct_defs = super::generate_structs_from_variants(
+        type_gen,
+        event.ty.id(),
+        |name| name.to_string(),
+        "Event",
+    );
     let event_structs = struct_defs.iter().map(|struct_def| {
         let pallet_name = &pallet.name;
         let event_struct = &struct_def.name;

--- a/codegen/src/api/mod.rs
+++ b/codegen/src/api/mod.rs
@@ -326,7 +326,7 @@ pub fn generate_structs_from_variants<'a, F>(
     error_message_type_name: &str,
 ) -> Vec<CompositeDef>
 where
-    F: Fn(&str) -> String,
+    F: Fn(&str) -> std::borrow::Cow<str>,
 {
     let ty = type_gen.resolve_type(type_id);
     if let scale_info::TypeDef::Variant(variant) = ty.type_def() {

--- a/codegen/src/api/mod.rs
+++ b/codegen/src/api/mod.rs
@@ -336,7 +336,7 @@ where
             .map(|var| {
                 let struct_name = variant_to_struct_name(var.name());
                 let fields = CompositeDefFields::from_scale_info_fields(
-                    &struct_name,
+                    struct_name.as_ref(),
                     var.fields(),
                     &[],
                     type_gen,

--- a/codegen/src/api/mod.rs
+++ b/codegen/src/api/mod.rs
@@ -319,19 +319,24 @@ impl RuntimeGenerator {
     }
 }
 
-pub fn generate_structs_from_variants<'a>(
+pub fn generate_structs_from_variants<'a, F>(
     type_gen: &'a TypeGenerator,
     type_id: u32,
+    variant_to_struct_name: F,
     error_message_type_name: &str,
-) -> Vec<CompositeDef> {
+) -> Vec<CompositeDef>
+where
+    F: Fn(&str) -> String,
+{
     let ty = type_gen.resolve_type(type_id);
     if let scale_info::TypeDef::Variant(variant) = ty.type_def() {
         variant
             .variants()
             .iter()
             .map(|var| {
+                let struct_name = variant_to_struct_name(var.name());
                 let fields = CompositeDefFields::from_scale_info_fields(
-                    var.name(),
+                    &struct_name,
                     var.fields(),
                     &[],
                     type_gen,

--- a/codegen/src/types/composite_def.rs
+++ b/codegen/src/types/composite_def.rs
@@ -22,7 +22,6 @@ use super::{
     TypeParameter,
     TypePath,
 };
-use heck::CamelCase as _;
 use proc_macro2::TokenStream;
 use proc_macro_error::abort_call_site;
 use quote::{
@@ -85,7 +84,7 @@ impl CompositeDef {
             }
         }
 
-        let name = format_ident!("{}", ident.to_camel_case());
+        let name = format_ident!("{}", ident);
 
         Self {
             name,

--- a/codegen/src/types/tests.rs
+++ b/codegen/src/types/tests.rs
@@ -853,3 +853,36 @@ fn modules() {
         .to_string()
     )
 }
+
+#[test]
+fn dont_force_struct_names_camel_case() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct AB;
+
+    let mut registry = Registry::new();
+    registry.register_type(&meta_type::<AB>());
+    let portable_types: PortableRegistry = registry.into();
+
+    let type_gen = TypeGenerator::new(
+        &portable_types,
+        "root",
+        Default::default(),
+        Default::default(),
+    );
+    let types = type_gen.generate_types_mod();
+    let tests_mod = get_mod(&types, MOD_PATH).unwrap();
+
+    assert_eq!(
+        tests_mod.into_token_stream().to_string(),
+        quote! {
+            pub mod tests {
+                use super::root;
+
+                #[derive(::subxt::codec::Encode, ::subxt::codec::Decode, Debug)]
+                pub struct AB;
+            }
+        }
+        .to_string()
+    )
+}

